lightning/src/chain/channelmonitor.rs

```diff
@@ -1088,6 +1084,38 @@ impl<Signer: WriteableEcdsaChannelSigner> Writeable for ChannelMonitorImpl<Signe
 	}
 }
 
+macro_rules! _process_events_body {
+	($self_opt: expr, $event_to_handle: expr, $handle_event: expr) => {
+		loop {
+			let (pending_events, repeated_events);
+			if let Some(us) = $self_opt {
+				let mut inner = us.inner.lock().unwrap();
+				if inner.pending_events_processor {
+					break;
+				}
+				inner.pending_events_processor = true;
+
+				pending_events = inner.pending_events.clone();
+				repeated_events = inner.get_repeated_events();
+			} else { break; }
+			let num_events = pending_events.len();
+
+			for event in pending_events.into_iter().chain(repeated_events.into_iter()) {
+				$event_to_handle = event;
+				$handle_event;
+			}
+
+			if let Some(us) = $self_opt {
+				let mut inner = us.inner.lock().unwrap();
+				inner.pending_events.drain(..num_events);
+				inner.pending_events_processor = false;
```

Comment:
Is the point of not holding the lock the entire time to allow new events to be added while processing?

Reply:
Hmm... I don't have a problem compiling while holding the lock using `cargo check -p lightning-background-processor --features=futures`. Is that what you meant by user code?

Reply:
Yeah, I see:

```
error: future cannot be sent between threads safely
   --> src/lib.rs:686:17
    |
686 |           runtime.spawn(async move {
    |  _______________________^
687 | |             process_events_async(
688 | |                 background_persister,
689 | |                 |e| background_event_handler.handle_event(e),
...   |
703 | |             });
704 | |         });
    | |_________^ future created by async block is not `Send`
    |
    = help: within `[async block@src/lib.rs:686:17: 704:4]`, the trait `std::marker::Send` is not implemented for `std::sync::RwLockReadGuard<'_, HashMap<lightning::chain::transaction::OutPoint, chainmonitor::MonitorHolder<InMemorySigner>>>`
note: required by a bound in `Runtime::spawn`
   --> /Users/jkczyz/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.29.1/src/runtime/runtime.rs:205:21
    |
205 |         F: Future + Send + 'static,
    |                     ^^^^ required by this bound in `Runtime::spawn`
```